File: app/src/main/java/org/openweatherapp/MainActivity.java
package org.openweatherapp;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import org.openweatherapp.adapters.CityListAdapter;
import org.openweatherapp.models.City;
import org.openweatherapp.viewmodels.CityViewModel;

public class MainActivity extends AppCompatActivity {

    private CityViewModel mCityViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView recyclerView = findViewById(R.id.recyclerview);

        final CityListAdapter adapter = new CityListAdapter(new CityListAdapter.CityDiff());
        recyclerView.setAdapter(adapter);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        RecyclerView.ItemDecoration itemDecoration =
                new DividerItemDecoration(this, DividerItemDecoration.VERTICAL);
        recyclerView.addItemDecoration(itemDecoration);

        mCityViewModel = new ViewModelProvider(this).get(CityViewModel.class);

        adapter.submitList(mCityViewModel.getCities());

        adapter.setOnItemClickListener(new CityListAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(View itemView, int position) {
                City city = adapter.getCurrentList().get(position);
                Intent intent = new Intent(MainActivity.this, WeatherActivity.class);
                intent.putExtra(WeatherActivity.KEY_CITY, city.getCity());
                startActivity(intent);
            }
        });
    }
}

File: app/src/main/java/org/openweatherapp/ui/cities/MainActivity.kt
package org.openweatherapp.ui.cities

import android.content.Intent
import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import org.openweatherapp.R
import org.openweatherapp.databinding.ActivityMainBinding
import org.openweatherapp.ui.weather.WeatherActivity

const val EXTRA_CITY_NAME = "org.openweatherapp.CITY_NAME"

class MainActivity: AppCompatActivity() {

    private val viewModel: CityViewModel by viewModels {
        CityViewModelFactory()
    }

    // derived from the name of the layout file, that is, activity_main + Binding
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        val adapter = CityAdapter(CityListener { cityName ->
            viewModel.onCityClicked(cityName)
        })

        binding.recyclerview.adapter = adapter

        adapter.submitList(viewModel.cities)

        binding.lifecycleOwner = this

        viewModel.navigateToCityDetail.observe(this, Observer { cityName ->
            cityName?.let {
                val intent = Intent(this, WeatherActivity::class.java).apply {
                    putExtra(EXTRA_CITY_NAME, cityName)
                }
                startActivity(intent)
                viewModel.onCityDetailNavigated()
            }
        })
    }
}